doc/remediation-templating.md

```diff
@@ -0,0 +1,342 @@
+   * As a content writer, it is possible to use XCCDF value from rules in Kubernetes remediation
+   * Add additional remediation state that shows "Needs-Review" for not set and no default values 
+     remediations
+
+## High-level overview
+
+The compliance operator allows an administrator to extend a pre-existing profile, 
+to enable, disable rules and set values that come from the ProfileBundle. When 
+this happens, a CR `TailoredProfile` will be generated. Currently the `ComplianceScan`
+will use the tailored rules and variables from `TailoredProfile` CR, however,
+the `ComplianceRemediation` will now take custom values from `TailoredProfile`.
+Here we are going to implement that feature.
+
+### ComplianceAsCode remediation rule format
+
+
+In ComplianceAsCode, when the SCAP remediation content is generated, normally we will 
+have fixed remediation. In this proposal, we are going to change the remediation rule
+formate for the Kubernetes to include a marked reference variable to reference related
+XCCDF values from rules. 
+
+The following example is parts of two YAML files from ComplianceAsCode content, 
+the first one is the rule, and the second is the remediation:
+```yaml
+documentation_complete: true
+title: 'Set SSH Idle Timeout Interval'
+...
+requires:
+    - sshd_set_keepalive
+
+ocil_clause: 'it is commented out or not configured properly'
+
+ocil: |-
+    Run the following command to see what the timeout interval is:
+    <pre>$ sudo grep ClientAliveInterval /etc/ssh/sshd_config</pre>
+    If properly configured, the output should be:
+    <pre>ClientAliveInterval {{{ xccdf_value("sshd_idle_timeout_value") }}}</pre>
+...
+```
+As showing in the rule YAML file above, an XCCDF value for SSH Idle Timeout Interval is set.
+
+In the remediation YAML file, we can modify that Marco to pass the name of the 
+XCCDF vairable and its value `sshd_idle_timeout_value`, and then he Marco will substitute the related 
+`ClientAliveInterval` values to `VAR_SSHD_IDLE_TIMEOUT_VALUE` as defined in the ocil:
+```yaml
+---
+# platform = multi_platform_ocp,multi_platform_rhcos
+# reboot = false
+# strategy = restrict
+# complexity = low
+# disruption = low
+{{{ kubernetes_sshd_set("ClientAliveInterval","sshd_idle_timeout_value") }}}
+```
+
+And the machine config object data will be something like below before it gets url-encoded.
+```jinja
+...
+PrintLastLog yes
+#TCPKeepAlive yes
+PermitUserEnvironment no
+Compression no
+ClientAliveInterval 600
+ClientAliveCountMax VAR_SSHD_IDLE_TIMEOUT_VALUE
+
+#UseDNS no
+#PidFile /var/run/sshd.pid
+#MaxStartups 10:30:100
+#PermitTunnel no
+...
+```
+Since the url encode doesn't encode ASCII alpha-numerical nor underscore characters, and
+we can later substitute the place holder after the url encoding from `VAR_SSHD_IDLE_TIMEOUT_VALUE`
+to `{{sshd_idle_timeout_value | findValue | urlquery}}`, so that it can be post-process them with the appropriate
```

Comment:
I don't follow the urlquery part of it. Is it for an already URL encoded value?

Reply:
Since the `MachineConfig` object is url-encoded and the value for "sshd_idle_timeout_value" is not, we need to encoded that value as well before we substitute it into the `MachineConfig` object.

Reply:
Ah, I see now.